api/ombpdf/semdom.py

```diff
@@ -0,0 +1,186 @@
+        if self.cursor == self.root:
+            self._push_child('sec')
+
+    def _add_text(self, text):
+        self.cursor.appendChild(self.document.createTextNode(text))
+
+    def _add_comment(self, text):
+        self.cursor.appendChild(self.document.createComment(text))
+
+    # Below are Writer methods.
+
+    def begin_document(self, doc):
+        self.document = minidom.parseString('<policy></policy>')
+        self.root = self.document.childNodes[0]
+        self.cursor = self.root
+
+    def end_document(self, doc):
+        pass
+
+    def begin_heading(self, heading):
+        while self.depth >= heading.level:
+            self._pop_child()
+        self._push_child('sec', {'title': ''})
+        self._push_child('heading')
+        self.muted = False
+
+    def end_heading(self, heading):
+        self._pop_child('heading')
+        self.muted = True
+
+    def begin_paragraph(self, p):
+        self.muted = False
+        self._ensure_section()
+        self._push_child('para')
+        self._push_child('content')
+
+    def end_paragraph(self, p):
+        self._pop_child('content')
+        self._pop_child('para')
+        self.muted = True
+
+    def begin_list(self, li):
+        self._ensure_section()
+        self._push_child('list', elinfo=ListInfo(self, li))
+
+    def end_list(self, li):
+        self._pop_child('list')
+
+    def begin_list_item(self, p):
+        self.elinfo[self.cursor].begin_list_item()
+        self.muted = False
+
+    def end_list_item(self, p):
+        self.elinfo[self.cursor].end_list_item()
+        self.muted = True
+
+    def create_footnote_citation(self, cit):
+        self._push_child('footnote_citation')
+        self._add_text(str(cit.number))
+        self.footnote_citations[cit.number] = self.cursor
+        self._pop_child('footnote_citation')
+
+    def begin_footnote_list(self, fl):
+        self.muted = True
+
+    def end_footnote_list(self, fl):
+        if self.footnote_citations:
+            unresolved = str(list(self.footnote_citations.keys()))
+            self._add_comment(
+                'Warning, unresolved footnote citations exist: '
+                f'{unresolved}'
+            )
+        self.muted = False
+
+    def create_footnote(self, f):
+        citation = self.footnote_citations.get(f.number)
+        if citation is not None:
+            del self.footnote_citations[f.number]
+            parent = citation.parentNode
+            if parent.nodeName == 'content':
+                parent = parent.parentNode
+            self._push_cursor(parent)
+            self._push_child('footnote', {
+                'emblem': str(f.number),
+                'marker': str(f.number),
+            })
+            self._add_text(f.text)
+            self._pop_child('footnote')
+            self._pop_cursor()
+        else:
+            self._add_comment(f'Uncited footnote #{f.number}: {f.text}')
+
+    def create_text(self, text):
+        text = text.replace('\n', '')
```

Comment:
What do you think about replacing newlines with spaces instead? Otherwise `some\nwords` will become `somewords`.

Reply:
Ah yeah gud point, I will bust that up.